### PR TITLE
Enhanced EventEmitter to pass data with an event. fixes #1979

### LINKED ISSUE #1979: Enhance eventEmitter to be able to send data with an event
It would be great if we could send data whenever we trigger an event within the player.

Just like jquery
https://api.jquery.com/trigger/
or node do
https://nodejs.org/api/events.html#events_emitter_emit_event_arg1_arg2

### PR DESCRIPTION
**Pull request by carpasse**
Extended trigger function to pass the data to the event handlers

### CODE REVIEW

**Comment by gkatsev** [comment]
Awesome stuff.

File: src/js/utils/events.js [hunk 1]
```diff
@@ -53,7 +53,7 @@ export function on(elem, type, fn){
           if (event.isImmediatePropagationStopped()) {
             break;
           } else {
-            handlersCopy[m].call(elem, event);
+            handlersCopy[m].apply(elem, [event].concat(evtData));
```
Comment (gkatsev) [inline src/js/utils/events.js]: I think you should be able to do something like
```js
handlersCopy[m].call(elem, event, ...evtData);
```

Comment (gkatsev) [inline src/js/utils/events.js]: Same as below.

File: src/js/component.js [hunk 2]
```diff
@@ -684,8 +684,8 @@ class Component {
    * @param  {Event|Object|String} event  A string (the type) or an event object with a type attribute
    * @return {Component}       self
    */
-  trigger(event) {
-    Events.trigger(this.el_, event);
+  trigger(event, ...data) {
```
Comment (gkatsev) [inline src/js/component.js]: I think we're thinking of requiring that data be an options object and just having the one extra parameter for data.

Comment (heff) [inline src/js/component.js]: Yeah, we're strictly following the Airbnb guide now which notes that.
https://github.com/airbnb/javascript#events

File: src/js/utils/events.js [hunk 3]
```diff
@@ -127,8 +127,9 @@ export function off(elem, type, fn) {
  * Trigger an event for an element
  * @param  {Element|Object}      elem  Element to trigger an event on
  * @param  {Event|Object|String} event A string (the type) or an event object with a type attribute
+ * @param  {...*} [evtData] you may pass additional data with the event
  */
-export function trigger(elem, event) {
+export function trigger(elem, event, ...evtData) {
   // Fetches element data and a reference to the parent (for bubbling).
```
Comment (carpasse) [inline src/js/utils/events.js]: Would it make sense to enforce the use of a data hash on Events.trigger as I did on component.trigger??

Comment (heff) [inline src/js/utils/events.js]: Yeah, that makes sense to me.

File: src/js/utils/events.js (hunk 1 above)
Comment (heff) [inline src/js/utils/events.js]: @carpasse did you want to update these to match the single hash arg or are you happy with it as is? Other than this question I think it's good to pull in.

Comment (carpasse) [inline src/js/utils/events.js]: @heff I did update them 3 days ago. I don't know why they are not shown here. 
Can you please take a look a the changes in https://github.com/videojs/video.js/compare/master...carpasse:feature/event-emitter-enhance

**Comment by carpasse** [comment]
@heff My bad. I did miss one 'rest' argument in the 'on' method dispatcher. It should be fixed now.

**Comment by heff** [comment]
No worries. Looks good to me. I'll merge in soon.

**Comment by heff** [comment]
Finally got this merged in. Thanks!

**Pull request closed without merging**